File: packages/vue-im-ui/src/shared/contracts/url-safety.ts
/**
 * Which URLs a chat surface may navigate to.
 *
 * Message content is written by other people. A link in it is a string an
 * attacker chose, so the only safe default is: render it as text unless it is
 * plainly a web address. `javascript:` and `data:` in an `href` execute in the
 * page's origin; `file:` and `blob:` read local state.
 *
 * The kit itself almost never navigates — components report the URL and the host
 * opens it. This is the rule both sides apply, so the kit's two anchors and the
 * host's own open-link handler cannot disagree.
 */

/** The only schemes a message link may carry. */
export const FLARE_SAFE_URL_PROTOCOLS: readonly string[] = ["http:", "https:"];

/**
 * The URL to navigate to, or null when the string is not a safe web address.
 *
 * A scheme-less string is read as https — that is what a user typing
 * `example.com` means — but only when it carries no scheme at all, so
 * `javascript:alert(1)` is never rescued into `https://javascript:alert(1)`.
 */
export function safeExternalUrl(raw: string | null | undefined): string | null {
  if (typeof raw !== "string") return null;
  // Strip only what a URL parser itself ignores — tab, newline, carriage return
  // — so `java\tscript:` normalizes to `javascript:` before the check. Interior
  // spaces must stay: removing them would turn a sentence (`just some text`)
  // into a valid host name.
  const trimmed = raw.trim().replace(/[\t\n\r]/g, "");
  if (!trimmed) return null;
  // A real URL carries no literal space. Refusing one keeps a sentence from
  // being read as an address, and keeps the four platforms identical by rule
  // rather than by whichever parser happens to be stricter: Dart's `Uri` would
  // percent-encode `just some text` into a valid host, JavaScript's `URL` throws.
  if (/\s/.test(trimmed)) return null;
  const candidate = hasScheme(trimmed) ? trimmed : `https://${trimmed}`;
  let parsed: URL;
  try {
    parsed = new URL(candidate);
  } catch {
    return null;
  }
  if (!FLARE_SAFE_URL_PROTOCOLS.includes(parsed.protocol)) return null;
  if (!parsed.hostname) return null;
  return parsed.href;
}

/** Whether the string is a web address this surface may open. */
export function isSafeExternalUrl(raw: string | null | undefined): boolean {
  return safeExternalUrl(raw) !== null;
}

/**
 * `scheme:` at the start, per RFC 3986 — letters, digits, `+`, `-`, `.`.
 *
 * A host and a port look the same to that grammar (`example.com:8443` is a legal
 * scheme followed by nothing), so a colon followed by digits is read as a port.
 * The worst that mis-reading can produce is an https URL to a host with an odd
 * name, which is harmless; reading a real scheme as a host is what must not
 * happen, and that is the direction this errs away from.
 */
function hasScheme(value: string): boolean {
  const match = /^[a-z][a-z0-9+.-]*:/i.exec(value);
  if (!match) return false;
  return !/^\d/.test(value.slice(match[0].length));
}

File: packages/flutter-im-ui/lib/src/models/url_safety.dart
/// 聊天界面可以跳转的 URL。
///
/// 消息内容是别人写的,里面的链接就是攻击者挑的字符串。`javascript:` 和 `data:`
/// 在 href 上会在页面来源里执行,`file:` / `blob:` 会读本地状态。所以默认只有一条:
/// 不是明摆着的网址就当纯文本。
///
/// kit 几乎从不自己跳转——组件把 URL 报给宿主,由宿主打开。这是两边共用的判定,
/// 免得 kit 的链接卡片和宿主的打开逻辑各有一套。
library;

/// 消息链接唯一允许的两个 scheme。
const List<String> flareSafeUrlSchemes = ['http', 'https'];

/// 可跳转时返回规范化后的 URL,否则返回 null。
///
/// 完全不带 scheme 的字符串按 https 读——用户打 `example.com` 就是这个意思——
/// 但只在真的一个 scheme 都没有时才补,`javascript:alert(1)` 绝不会被救成
/// `https://javascript:alert(1)`。
String? safeExternalUrl(String? raw) {
  if (raw == null) return null;
  // 只剥 URL 解析器本来就会忽略的制表符与换行:`java\tscript:` 会规范成
  // `javascript:`,先剥掉才判得准。**不能连空格一起剥**——那会把一句话
  // (`just some text`)捏成一个合法主机名。
  final trimmed = raw.trim().replaceAll(RegExp(r'[\t\n\r]'), '');
  if (trimmed.isEmpty) return null;
  // 真正的 URL 里没有裸空格。拒掉它,一是不让一句话被当成网址,二是让四端靠
  // 同一条规则对齐,而不是靠谁的解析器更严:Dart 的 Uri 会把 `just some text`
  // 百分号编码成一个合法主机,JS 的 URL 会直接抛。
  if (RegExp(r'\s').hasMatch(trimmed)) return null;
  final candidate = _hasScheme(trimmed) ? trimmed : 'https://$trimmed';
  final parsed = Uri.tryParse(candidate);
  if (parsed == null) return null;
  if (!flareSafeUrlSchemes.contains(parsed.scheme.toLowerCase())) return null;
  if (parsed.host.isEmpty) return null;
  return parsed.toString();
}

/// 这个字符串是不是本界面可以打开的网址。
bool isSafeExternalUrl(String? raw) => safeExternalUrl(raw) != null;

/// 开头是 `scheme:`(RFC 3986:字母、数字、`+`、`-`、`.`)。
///
/// 主机加端口在这套语法里长得一样(`example.com:8443` 也是合法 scheme),所以冒号
/// 后面跟数字就当端口。认错的最坏结果是一个主机名古怪的 https URL,无害;把真
/// scheme 认成主机才是不能发生的,这里的偏向就是躲开那一侧。
bool _hasScheme(String value) {
  final match = RegExp(r'^[a-z][a-z0-9+.\-]*:', caseSensitive: false).firstMatch(value);
  if (match == null) return false;
  return !RegExp(r'^\d').hasMatch(value.substring(match.end));
}
